### DISCUSSION TITLE
Custom compilation failed error: 'lv_font_montserrat_14' was not declared in this scope

### DISCUSSION THREAD

**Discussion by pp1wvc** [HASwitchPlate/openHASP#450] [Q&A]
I found a failure when compiling openhasp with vscod and gitpod, the failures are identical:

Compiling .pio/build/freetouchdeck_4MB/src/hasp/hasp_style.cpp.o
Compiling .pio/build/freetouchdeck_4MB/src/hasp/hasp_task.cpp.o
In file included from include/lv_conf_v7.h:1068,
                 from include/lv_conf.h:5,
                 from src/hasplib.h:23,
                 from src/mqtt/hasp_mqtt.h:8,
                 from include/hasp_conf.h:274,
                 from src/hasplib.h:20,
                 from src/hasp/hasp.cpp:4:
src/hasp/hasp.cpp: In function 'void haspSetup()':
.pio/libdeps/freetouchdeck_4MB/lvgl/src/lv_conf_internal.h:1236:51: error: 'lv_font_montserrat_14' was not declared in this scope
 #    define  LV_THEME_DEFAULT_FONT_SMALL         &lv_font_montserrat_14
                                                   ^~~~~~~~~~~~~~~~~~~~~
src/hasp/hasp.cpp:582:20: note: in expansion of macro 'LV_THEME_DEFAULT_FONT_SMALL'
     haspFonts[0] = LV_THEME_DEFAULT_FONT_SMALL;
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
.pio/libdeps/freetouchdeck_4MB/lvgl/src/lv_conf_internal.h:1236:51: note: suggested alternative: 'lv_font_user_data_t'
 #    define  LV_THEME_DEFAULT_FONT_SMALL         &lv_font_montserrat_14
                                                   ^~~~~~~~~~~~~~~~~~~~~
src/hasp/hasp.cpp:582:20: note: in expansion of macro 'LV_THEME_DEFAULT_FONT_SMALL'
     haspFonts[0] = LV_THEME_DEFAULT_FONT_SMALL;
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
*** [.pio/build/freetouchdeck_4MB/src/hasp/hasp.cpp.o] Error 1
src/hasp/hasp_font.cpp: In function 'lv_font_t* font_add_to_list(const char*)':
src/hasp/hasp_font.cpp:171:32: warning: ISO C++ forbids converting a string constant to 'char*' [-Wwrite-strings]
     char* ext[] = {"ttf", "otf"};
                                ^
src/hasp/hasp_font.cpp:171:32: warning: ISO C++ forbids converting a string constant to 'char*' [-Wwrite-strings]
======================================== [FAILED] Took 97.34 seconds ========================================

Environment        Status    Duration
-----------------  --------  ------------
freetouchdeck_4MB  FAILED    00:01:37.341
=================================== 1 failed, 0 succeeded in 00:01:37.341 ===================================
gitpod /workspace/openHASP $ 

By VsCode:

ompiling .pio\build\freetouchdeck_4MB\src\hasp\hasp_mem.cpp.o
In file included from include/lv_conf_v7.h:1068,
                 from include/lv_conf.h:5,
                 from src/hasplib.h:23,
                 from src/mqtt/hasp_mqtt.h:8,
                 from include/hasp_conf.h:274,
                 from src/hasplib.h:20,
                 from src/hasp/hasp.cpp:4:
src/hasp/hasp.cpp: In function 'void haspSetup()':
.pio/libdeps/freetouchdeck_4MB/lvgl/src/lv_conf_internal.h:1236:51: error: 'lv_font_montserrat_14' was not declared in this 
scope
 #    define  LV_THEME_DEFAULT_FONT_SMALL         &lv_font_montserrat_14
                                                   ^~~~~~~~~~~~~~~~~~~~~
src/hasp/hasp.cpp:582:20: note: in expansion of macro 'LV_THEME_DEFAULT_FONT_SMALL'
     haspFonts[0] = LV_THEME_DEFAULT_FONT_SMALL;
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
.pio/libdeps/freetouchdeck_4MB/lvgl/src/lv_conf_internal.h:1236:51: note: suggested alternative: 'lv_font_user_data_t'
 #    define  LV_THEME_DEFAULT_FONT_SMALL         &lv_font_montserrat_14
                                                   ^~~~~~~~~~~~~~~~~~~~~
src/hasp/hasp.cpp:582:20: note: in expansion of macro 'LV_THEME_DEFAULT_FONT_SMALL'
     haspFonts[0] = LV_THEME_DEFAULT_FONT_SMALL;
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
*** [.pio\build\freetouchdeck_4MB\src\hasp\hasp.cpp.o] Error 1
src/hasp/hasp_font.cpp: In function 'lv_font_t* font_add_to_list(const char*)':
src/hasp/hasp_font.cpp:171:32: warning: ISO C++ forbids converting a string constant to 'char*' [-Wwrite-strings]
     char* ext[] = {"ttf", "otf"};
                                ^
src/hasp/hasp_font.cpp:171:32: warning: ISO C++ forbids converting a string constant to 'char*' [-Wwrite-strings]
=============================================== [FAILED] Took 69.29 seconds ===============================================

Environment        Status    Duration
-----------------  --------  ------------
freetouchdeck_4MB  FAILED    00:01:09.285
========================================== 1 failed, 0 succeeded in 00:01:09.285 ========================================== 

 *  The terminal process "C:\Users\user\.platformio\penv\Scripts\platformio.exe 'run'" terminated with exit code: 1. 
 *  Terminal will be reused by tasks, press any key to close it. 
 * 'lv_font_montserrat_14' was not declared in this
 * #    define  LV_THEME_DEFAULT_FONT_SMALL         &lv_font_montserrat_14
 * lv_conf_internal.h

these are the changes:

; USAGE: Copy this file and rename it to platformio_override.ini
;
; The platformio_override.ini file is not overwritten or monitored by git
; ONLY edit platformio_override.ini to make local changes to the parameters

[platformio]
extra_configs =
    ; Uncomment or edit the lines to show more User Setups in the PIO sidebar
    ; user_setups/darwin_sdl/*.ini
     user_setups/esp32/*.ini
    ; user_setups/esp32s2/*.ini
    ; user_setups/esp32s3/*.ini
    ; user_setups/linux_sdl/*.ini
    ; user_setups/stm32f4xx/*.ini
    ; user_setups/win32/*.ini

[override]
; -- Hasp config options --------------------------------------
build_flags =
; -- Uncomment the next line to use the file include/user_config_override.h settings
    -DUSE_CONFIG_OVERRIDE

;region -- Default Build Environments : Used when Build All ---
extra_default_envs =
    ; Uncomment specific environments or create extra:
    ; az-touch-mod-esp32_ili9341_4MB
    ; az-touch-mod-esp32_ili9341_8MB
    ; d1-mini-esp32_ili9341
    ; d1-r32-unoshield_ili9341_adc
    ; d1-r32-unoshield_ili9486_adc
    ; d1-r32-waveshare_ili9486
    ; esp32-2432s028r_4MB
    ; esp32-3248s035c_4MB
    ; esp32-3248s035c_4MB_lovyan
    ; esp32-3248s035r_4MB
    ; esp32-one_ili9486
    ; esp32-one_st7796
    ; esp32-s3-4827s043c_16MB
    ; esp32-s3-4827s043r_16MB
    ; esp32-s3-8048s043c_16MB
    ; esp32-touchdown
     freetouchdeck_4MB
    ; freetouchdeck_8MB
    ; freetouchdeck-s3_8MB
    ; gs-t3e_16MB
    ; huzzah32-featherwing-24
    ; huzzah32-featherwing-35
    ; lanbon_l8
    ; lanbon_l8_eth
    ; lilygo-lily-pi_ili9481
    ; lilygo-lily-pi_st7796
    ; lolin-d32-pro_ili9341
    ; m5stack-core2
    ; m5stack-tough
    ; makerfabs-s3-tft35-spi
    ; makerfabs-tft35-cap
    ; makerfabs-tft-s2_ili9488
    ; nodemcu32s-raspi
    ; s2-mini-esp32s2_ili9341
    ; ttgo_esp32_poe-ili9341
    ; ttgo-lilygo-pi_ili9481
    ; ttgo-lilygo-pi_st7796
    ; ttgo-t7-v1.5_ili9341_16MB
    ; ttgo-t7-v1.5_ili9341_4MB
    ; ttgo-esp32_tdisplay_v1
    ; wt32-sc01_16MB
    ; wt32-sc01_4MB
    ; wt32-sc01-plus_16MB
    ; wt32-sc01-plus_8MB
    ; wt-86-32-3zw1
    ; yeacreate-nscreen32
;endregion

;region -- Define your local COM ports for each environment ---
[env:d1-mini-esp32_ili9341]
monitor_port = COM6       ; Change to the correct port
upload_port = ${env:d1-mini-esp32_ili9341.monitor_port}

[env:esp32dev-mrb3511]
monitor_port = COM3       ; Change to the correct port
upload_port = ${env:esp32dev-mrb3511.monitor_port}

[env:az-touch-mod-esp32_ili9341_4MB]
upload_port = COM16 
monitor_port = COM16

[env:az-touch-mod-esp32_ili9341_8MB]
upload_port = COM16 
monitor_port = COM16

[env:d1-r32-unoshield_ili9486_adc]
monitor_port = COM5 
upload_port = COM5  

[env:d1-r32-unoshield_ili9341_adc]
;upload_protocol = espota  ; Use ArduinoOTA after flashing over serial
;upload_port = 192.168.0.4 ; IP of the ESP
;upload_flags = --port=3232 ; --auth=haspadmin   ; OTA password
monitor_port = COM5 
upload_port = COM5

[env:esp32-one_st7796]
upload_port = COM6      ; Change to the correct port
monitor_port = COM6     ; Change to the correct port

[env:lolin-d32-pro_ili9341]
monitor_port = COM5       ; Change to the correct port
upload_port = ${env:lolin-d32-pro_ili9341.monitor_port}
;endregion

[env:ttgo_esp32_poe-ili9341]
monitor_port = COM9       ; Change to the correct port
;upload_port = ${env:ttgo_esp32_poe-ili9341.monitor_port}
;upload_port = 192.168.4.5 ; IP of the ESP
;upload_protocol = espota  ; Use ArduinoOTA after flashing over serial
upload_flags = --port=3232 ; --auth=haspadmin   ; OTA password

;region -- Custom Environment configuration example -----------------
[env:my_custom_build]
extends = arduino_esp32_v2
board = nodemcu-32s

upload_port = COM12     ; To change the port, use platform_override.ini
monitor_port = COM12    ; To change the port, use platform_override.ini

board_build.partitions = user_setups/esp32/partitions_4MB.csv ; default.csv

debug_tool = esp-prog
debug_init_break = tbreak setup
build_flags =
    ${env.build_flags}
    ${esp32.build_flags}
;region -- TFT_eSPI build options -----------------------------------
    ${lcd.lolin24}
    ${esp32.vspi}        ; Use VSPI hardware SPI bus
    -D TFT_DC=5
    -D TFT_CS=26 
    -D TFT_RST=-1        ; RST
    -D TFT_BCKL=22       ; None, configurable via web UI (e.g. 21)
    -D TOUCH_CS=17       ; (can also be 22 or 16)
; -- Options ----------------------------------------
    -D HASP_USE_TELNET=1
;endregion

;endregion

/* MIT License - Copyright (c) 2019-2022 Francis Van Roie
   For full license information read the LICENSE file in the project folder */

/***************************************************
// This file contains the default settings that are
// burned into the compiled firmware.
//
// These default settings can be changed at runtime
//
// To use: Save a copy as user_config_override.h
***************************************************/
#ifndef HASP_USER_CONFIG_OVERRIDE_H
#define HASP_USER_CONFIG_OVERRIDE_H

#define SERIAL_SPEED 115200

/***************************************************
          WiFi Settings
 **************************************************/
#define WIFI_SSID ""
#define WIFI_PASSWORD ""

/***************************************************
          Http Server Settings
 **************************************************/
#define HTTP_USERNAME ""
#define HTTP_PASSWORD ""

/***************************************************
          MQTT Client Settings
 **************************************************/
#define MQTT_HOSTNAME ""
#define MQTT_PORT 1883
#define MQTT_USERNAME ""
#define MQTT_PASSWORD ""
#define MQTT_PREFIX "hasp"
#define MQTT_NODENAME "plate10"
#define MQTT_GROUPNAME "plates"

#define MQTT_TELEPERIOD 60000
#define MQTT_STATEPERIOD 300000

/***************************************************
 *        OTA Settings
 **************************************************/
//#define HASP_USE_ARDUINOOTA 1                       // Enable the Arduino OTA service
#define ARDUINOOTA_PORT 3232
#define ARDUINOOTA_PASSWORD ""
#define OTA_URL ""

/***************************************************
 *        Syslog Settings
 **************************************************/
#define HASP_USE_SYSLOG 0
#define SYSLOG_SERVER ""
#define SYSLOG_PORT 514
#define APP_NAME "HASP"

/***************************************************
 *        Timezone Settings
 **************************************************/
#define NTPSERVER1 "pool.ntp.org"
#define NTPSERVER2 "time.nist.gov"
#define NTPSERVER3 "time.google.com"
#define MYTZ                                                                                                           \
    "CET-1CEST,M3.5.0,M10.5.0/3" // A full list with possible timezones can be found here
                                 // https://gist.github.com/alwynallan/24d96091655391107939

/***************************************************
 *        Interface Language Settings
 **************************************************/
// #define HASP_LANGUAGE en_US // English
// #define HASP_LANGUAGE es_ES                       // Spanish
// #define HASP_LANGUAGE fr_FR                       // French
// #define HASP_LANGUAGE hu_HU                       // Hungarian
// #define HASP_LANGUAGE nl_NL                       // Dutch
 #define HASP_LANGUAGE pt_BR                       // Brazilian Portuguese
// #define HASP_LANGUAGE pt_PT                       // Portuguese
// #define HASP_LANGUAGE ro_RO                       // Romanian

/***************************************************
 *        Web interface coloring
 **************************************************/
// Light theme (default)
#define D_HTTP_COLOR_TEXT "#000"               // Global text color - Black
#define D_HTTP_COLOR_BACKGROUND "#fff"         // Global background color - White
#define D_HTTP_COLOR_INPUT_TEXT "#000"         // Input text color - Black
#define D_HTTP_COLOR_INPUT "#fff"              // Input background color - White
#define D_HTTP_COLOR_INPUT_WARNING "#f00"      // Input warning border color - Red
#define D_HTTP_COLOR_BUTTON_TEXT "#fff"        // Button text color - White
#define D_HTTP_COLOR_BUTTON "#1fa3ec"          // Button color - Vivid blue
#define D_HTTP_COLOR_BUTTON_HOVER "#0083cc"    // Button color - Olympic blue
#define D_HTTP_COLOR_BUTTON_RESET "#f00"       // Restart/Reset button color - red
#define D_HTTP_COLOR_BUTTON_RESET_HOVER "#b00" // Restart/Reset button color - Dark red
#define D_HTTP_COLOR_GROUP "#f3f3f3"           // Group container background color
#define D_HTTP_COLOR_GROUP_TEXT "#000"         // Group container text color - black
#define D_HTTP_COLOR_FOOTER_TEXT "#0083cc"     // Footer text color - Olympic blue

/*
// Dark theme
#define D_HTTP_COLOR_TEXT               "#eaeaea"    // Global text color - Very light gray
#define D_HTTP_COLOR_BACKGROUND         "#252525"    // Global background color - Very dark gray (mostly black)
#define D_HTTP_COLOR_INPUT_TEXT         "#000"       // Input text color - Black
#define D_HTTP_COLOR_INPUT              "#ddd"       // Input background color - Very light gray
#define D_HTTP_COLOR_INPUT_WARNING      "#ff5661"    // Warning text color - Brick Red
#define D_HTTP_COLOR_BUTTON_TEXT        "#faffff"    // Button text color - Very pale (mostly white) cyan
#define D_HTTP_COLOR_BUTTON             "#1fa3ec"    // Button color - Vivid blue
#define D_HTTP_COLOR_BUTTON_HOVER       "#0083cc"    // Button color - Olympic Blue
#define D_HTTP_COLOR_BUTTON_RESET       "#d43535"    // Restart/Reset/Delete button color - Strong red
#define D_HTTP_COLOR_BUTTON_RESET_HOVER "#b00"       // Restart/Reset button color - Dark red
#define D_HTTP_COLOR_GROUP              "#444"       // Group container background color - Dark gray
#define D_HTTP_COLOR_GROUP_TEXT         "#fff"       // Group container text color - white
#define D_HTTP_COLOR_FOOTER_TEXT        "#1fa3ec"    // Footer text color - Vivid blue
*/

/***************************************************
 *        Font Settings
 **************************************************/
// #define HASP_FONT_1 robotocondensed_regular_16_latin1 // Use available fonts from src/fonts directory
// #define HASP_FONT_2 robotocondensed_regular_22_latin1
// #define HASP_FONT_3 robotocondensed_regular_40_latin1
// #define HASP_FONT_4 robotocondensed_regular_48_latin1
// #define HASP_FONT_5 robotocondensed_regular_12_latin1

// #define ROBOTOCONDENSED_REGULAR_16_LATIN1 1           // Enable the selected fonts
// #define ROBOTOCONDENSED_REGULAR_22_LATIN1 1
// #define ROBOTOCONDENSED_REGULAR_40_LATIN1 1
// #define ROBOTOCONDENSED_REGULAR_48_LATIN1 1
// #define ROBOTOCONDENSED_REGULAR_12_LATIN1 1

// #define HASP_FONT_SIZE_1 16                           // Define used font sizes
// #define HASP_FONT_SIZE_2 22
// #define HASP_FONT_SIZE_3 40
// #define HASP_FONT_SIZE_4 48
// #define HASP_FONT_SIZE_5 12

#define LV_FONT_DEFAULT &HASP_FONT_1

/***************************************************
 *        GPIO Settings
 **************************************************/
//#define HASP_GPIO_TEMPLATE "[197658,263456,329249,655628,655886,656155,0,0]"  // Lanbon L8 3-gang GPIO config
//#define HASP_GPIO_TEMPLATE "[3214348,197658,263456,329249,94699520,0,0,0]" // Lanbon L8 Dimmer GPIO config

/***************************************************
 *        Other Settings
 **************************************************/
//#define HASP_USE_MDNS 0                             // Disable MDNS
//#define HASP_USE_CUSTOM 1                           // Enable compilation of custom code from /src/custom
#define HASP_USE_HA                                 // Enable Home Assistant auto-discovery
//#define HASP_START_CONSOLE 0                        // Disable starting of serial console at boot
//#define HASP_START_TELNET 0                         // Disable starting of telnet service at boot
//#define HASP_START_HTTP 0                           // Disable starting of web interface at boot
//#define HASP_START_FTP 0                            // Disable starting of ftp server at boot
//#define LV_MEM_SIZE (64 * 1024U)                    // 64KiB of lvgl memory (default 48)
//#define LV_VDB_SIZE (32 * 1024U)                    // 32KiB of lvgl draw buffer (default 32)
//#define HASP_DEBUG_OBJ_TREE                         // Output all objects to the log on page changes
//#define HASP_LOG_LEVEL LOG_LEVEL_VERBOSE            // LOG_LEVEL_* can be DEBUG, VERBOSE, TRACE, INFO, WARNING, ERROR, CRITICAL, ALERT, FATAL, SILENT
//#define HASP_LOG_TASKS                              // Also log the Taskname and watermark of ESP32 tasks

#endif // HASP_USER_CONFIG_OVERRIDE_H

What could be wrong?

**Comment by pp1wvc** [comment]
Hello, anyone? Any idea? Any help please?

**Reply by htvekov** [reply]
Try and remark this line out
`#define LV_FONT_DEFAULT &HASP_FONT_1`

**Reply by pp1wvc** [reply]
> Tente e comente esta linha `#define LV_FONT_DEFAULT &HASP_FONT_1`

Hello htvekov
I will try your suggestion, thank you.

**Reply by pp1wvc** [reply]
> Try and remark this line out `#define LV_FONT_DEFAULT &HASP_FONT_1`

Compiling .pio\build\freetouchdeck_4MB\src\drv\touch\touch_driver_gt911.cpp.o
Compiling .pio\build\freetouchdeck_4MB\src\hasp_gui.cpp.o
In file included from src/hasp_gui.cpp:12:
src/drv/touch/touch_driver.h:54:2: warning: #warning Building for TFT_eSPI XPT2046 [-Wcpp]
 #warning Building for TFT_eSPI XPT2046
  ^~~~~~~
In file included from src/drv/touch/touch_driver.h:56,
                 from src/hasp_gui.cpp:12:
src/drv/touch/touch_driver_tftespi.h:7:16: warning: extra tokens at end of #ifdef directive
 #ifdef ARDUINO && defined(USER_SETUP_LOADED)
                ^~
Linking .pio\build\freetouchdeck_4MB\firmware.elf
c:/users/user/.platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: .pio/build/freetouchdeck_4MB/src/mqtt/hasp_mqtt_ha.cpp.o:(.literal._Z22mqtt_ha_add_device_idsRN11ArduinoJson7V6210PD12JsonDocumentE+0x24): undefined reference to `mqttNodeTopic'
collect2.exe: error: ld returned 1 exit status
*** [.pio\build\freetouchdeck_4MB\firmware.elf] Error 1
=============================================== [FAILED] Took 46.77 seconds ===============================================
Environment        Status    Duration
-----------------  --------  ------------
freetouchdeck_4MB  FAILED    00:00:46.774
========================================== 1 failed, 0 succeeded in 00:00:46.774 ==========================================
 *  The terminal process "C:\Users\user\.platformio\penv\Scripts\platformio.exe 'run'" terminated with exit code: 1. 
 *  Terminal will be reused by tasks, press any key to close it. 

Failed again, OpenHasp Custom is not for amateurs, I can't find a solution.
Some help?

**Reply by htvekov** [reply]
Sorry, not familiar with freetouchdesk builds. 
Try and build  gs-t3e_16MB instead, still leaving out the `#define LV_FONT_DEFAULT &HASP_FONT_1` line.
I've recently build this T3E environment using Github online successfully just by removing that specific line only.

**Comment by bensuffolk** [comment]
> Try and remark this line out
> `#define LV_FONT_DEFAULT &HASP_FONT_1`

That worked for me.

**Comment by linuxbaba** [comment]
In lv_conf_v7.h about line 651
#define LV_FONT_MONTSERRAT_14    1
